### PR TITLE
[SPARK-13454] [SQL] Allow users to drop a table with a name starting with an underscore.

### PR DESCRIPTION
**Pull request by yhuai**
## What changes were proposed in this pull request?

This change adds a workaround to allow users to drop a table with a name starting with an underscore. Without this patch, we can create such a table, but we cannot drop it. The reason is that Hive's parser unquote an quoted identifier (see https://github.com/apache/hive/blob/release-1.2.1/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveLexer.g#L453). So, when we issue a drop table command to Hive, a table name starting with an underscore is actually not quoted. Then, Hive will complain about it because it does not support a table name starting with an underscore without using backticks (underscores are allowed as long as it is not the first char though).
## How was this patch tested?

Add a test to make sure we can drop a table with a name starting with an underscore.

https://issues.apache.org/jira/browse/SPARK-13454

### CODE REVIEW

**Comment by yhuai** [comment]
This PR is just for branch 1.6. Master does not have this problem.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala [hunk 1]
```diff
@@ -542,7 +542,14 @@ https://cwiki.apache.org/confluence/display/Hive/Enhanced+Aggregation%2C+Cube%2C
     case Token("TOK_DROPTABLE",
            Token("TOK_TABNAME", tableNameParts) ::
            ifExists) =>
-      val tableName = tableNameParts.map { case Token(p, Nil) => p }.mkString(".")
+      // Hive's parser will unquote an identifier (see the rule of QuotedIdentifier in
+      // HiveLexer.g of Hive 1.2.1). So, we will quote a table name part if it starts with
+      // underscore (_). Please note that although QuotedIdentifier rule allows backticks
+      // appearing in an identifier, Hive does not actually allow such an identifier be
+      // a table name. So, we
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala]: the comment is incomplete...

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala [hunk 2]
```diff
@@ -542,7 +542,14 @@ https://cwiki.apache.org/confluence/display/Hive/Enhanced+Aggregation%2C+Cube%2C
...
+      val tableName = tableNameParts.map {
+        case Token(p, Nil) => if (p.startsWith("_")) s"`$p`" else p
+      }.mkString(".")
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala]: `DropTable` also tries to unregister temporary tables of the same name using:

```scala
hiveContext.catalog.unregisterTable(TableIdentifier(tableName))
````

Now the `tableName` argument of `DropTable` is quoted when it's started with `_`, but `TableIdentifier` doesn't parse and unquote the table name. Thus the temporary table can't be dropped in this case.

Another problem of the above line is that, `DropTable.tableName` can be the form of `db.table`, but we throw the whole string into `TableIdentifier` without splitting them. But that's another bug.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 3]
```diff
@@ -47,6 +47,17 @@ class HiveMetastoreCatalogSuite extends SparkFunSuite with TestHiveSingleton {
...
+  test("SPARK-13454: drop a table with a name starting with underscore") {
+    hiveContext.range(10).write.saveAsTable("_spark13454")
+    hiveContext.sql("drop table `_spark13454`")
+
+    hiveContext.range(10).write.saveAsTable("_spark13454")
+    hiveContext.sql("drop table default.`_spark13454`")
+
+    hiveContext.range(10).write.saveAsTable("spark13454")
+    hiveContext.sql("drop table spark13454")
+  }
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: Let's add test case that

1. drops temporary tables
2. drops persistent and temporary tables with the same name

**Comment by yhuai** [comment]
OK I have addressed the comments. I moved the fix to DropTable.

**Comment by yhuai** [comment]
Merging to 1.6

**Pull request closed without merging**